Работа с инвентарём

Limuranius opened the pull request:
- Обучил нейросеть по распознаванию цифр в инвентаре
- Объединил её с распознавателем аспектов в инвентаре
- Доработал код с крафтом аспектов. Теперь при выставлении аспектов на поле сначала проверяется их количество, и если кол-во какого-то аспекта 0, то он создаётся

Числа распознаются с точностью где-то 95%, так что использовать такой вариант можно, но на свой страх и риск. Если распознанное число будет больше истинного, то есть риск, что аспекты кончатся, а система об этом даже не узнает. Поэтому как альтернатива в коде ThaumInteractor.py можно раскомментировать эти строчки, чтобы проставить всем аспектам количество 0, тогда не будет переоценки количества
![image](https://github.com/user-attachments/assets/40912f48-3e13-48f4-b9e7-de8c8cd01263)

Limuranius committed:
Небольшой рефакторинг кода

- Разбил длинный код на подмодули
- Подправил импорты
- Расставил пробелы и переносы строк

Limuranius committed:
Убрана необходимость в roboflow

- Залиты веса распознавания аспектов в формате onnx
- Написан код для запуска onnx модели

Limuranius committed:
Исправлены пути при создании exe

Limuranius committed:
Подправил лишние переносы

Limuranius committed:
Логика для распознавания инвентаря

- Разбил код для работы с моделями на распознавание объектов и классификацию
- Переделал класс Neurolink, чтобы он работал со всеми моделями
- Добавил распознавание чисел

Limuranius committed:
Исправлены небольшие ошибки

- ошибки с импортами
- ошибки с размерностью данных
- размеры при вырезании цифр

Limuranius committed:
Обучение классификатора цифр

Limuranius committed:
В GUI добавлен функционал инвентаря (создание и кол-во аспектов)

Метод mixAspect
- Добавил учёт количества аспектов. Если кол-во достигает нуля, то аспекты крафтятся
- Убрал исключение о создании базовых аспектов, но добавил исключение, когда базовые аспекты кончились

Метод updateAvailableAspectsInInventory
- Добавил подсчёт количества аспектов
- Исправил (надеюсь) ошибку с неправильным подсчётом пройденных страниц

takeAspect
- Уменьшение кол-ва аспекта на 1 и проверка, если он кончился

Limuranius committed:
Сделал классификацию цифр точнее

- Изменил архитектуру нейросети
- Добавил 5 пикселей паддинга изображениям

Limuranius committed:
Исправлен баг, когда кол-во аспектов всегда 0

Limuranius committed:
Исправлен баг, когда кол-во аспектов всегда 0

SergTyapkin commented on src/controllers/ThaumInteractor.py:
```diff
@@ -344,46 +294,78 @@ def scrollToAspect(self, aspect: Aspect) -> (int, int):
-        logging.info(f"Mixing aspect {aspect}...")
+        aspect.count -= 1
+
+    def mixAspect(self, aspect: Aspect, useShift=True, times=5) -> None:
```
Зачем по умолчанию создавать аспект 5 раз?

Limuranius replied:
В долгосрочной перспективе, когда надо будет создавать много аспектов, это быстрее. Дополнительно тратится время на перелистывание страниц. Если делать по одному аспекту, то он будет листать страницы каждый раз, когда понадобится сделать новый, а так будет небольшой буфер в 5 аспектов.

SergTyapkin commented:
Такое ощущение, что у тебя в requirements.txt попали все модули из глобальной области на локальной машине. Почисть ненужные, пожалуйста. 
Вот твои:
![image](https://github.com/user-attachments/assets/c5d5e508-b49a-4d5d-983f-5a3ef3c0c202)

И посмотри, как их мало в исходной версии в репе:
![image](https://github.com/user-attachments/assets/0a941a29-8978-43f9-9cbc-81f8faa628ca)

Limuranius committed:
Убрал классификатор цифр, улучшил object detection

- Убрал использование самописного классификатора цифр, теперь все цифры распознаются через object detection
- Переделал преобразование входных изображений к квадратным. Теперь на вход нейросетям подаются изображения с сохранённым aspect ratio (ранее они растягивались по координатам x и y, до нужного размера)

Limuranius commented:
В requirements.txt ошибки нет. Там просто onnxruntime за собой тянет ещё штук 7 других математических библиотек

Limuranius committed:
Укорочен requirements.txt

SergTyapkin committed:
Merge branch 'master' into inventory-recognition
